Feature/task estimation in scrum

Isaac811 opened the pull request.

Isaac811 committed:
Task-Estimation-added-topic-headers

Isaac811 committed:
Task-Estimation-added-topic-headers

Isaac811 committed:
Delete Task-Estimation-added-topic-headers.txt

Isaac811 committed:
Task-Estimation-added-Content

Isaac811 committed:
Task-Estimation-Topics

Isaac811 committed:
Update QA Handbook.md

Paddyjh reviewed (changes requested):
This Pr is adding content for the continuous integration topic but it is on the task estimation in scrum branch. I this might have been a mistake @Isaac811

Rosscrowley commented:
> This Pr is adding content for the continuous integration topic but it is on the task estimation in scrum branch. I this might have been a mistake @Isaac811

Hey Paddy, I reset the branch to the previous stable commit that you had @Isaac811, so that I can work on my section.

Rosscrowley commented on Handbook/QA Handbook.md:
```diff
@@ -2,7 +2,20 @@
+What is Task Estimation and why it is one of the key features?
+	In the initial stage of software development, the demand side usually has clear requirements and not very clear expectations for the final software product, and software quality refers to the degree of conformity with the final product. Specifically, software quality refers to the degree to which software conforms to clearly stated functional and performance requirements, development criteria, a ...
+
+Use an example to show these three aspects: 
```
Hey @Isaac811 and @Paddyjh, Do you think we could talk about more specific topics related to task estimation in scrum. The current topics may be to broad and focus more on the scrum process rather than task estimations in scrum.
What do we think about the following topics instead:  
- Estimation Techniques
- Challenges and considerations when accurately estimating task
- Adjusting Estimates During the Sprint
- Team Velocity and calculating Estimation

Paddyjh replied:
Hey, I think this makes a lot more sense, although I would probably adjust the topic of  "Adjusting Estimates During the Sprint" to something like "Regular Estimation Reviews and Adjustments" so we can talk more about incorporating feedback etc to improve the process.

What do we think ?

Isaac811 commented:
Seems all make sense.

Best regards

Jiayu Liu


Patrick Haughey ***@***.***>于2024年3月23日 周六00:16写道：

> ***@***.**** requested changes on this pull request.
>
> This Pr is adding content for the continuous integration topic but it is
> on the task estimation in scrum branch. I this might have been a mistake
> @Isaac811 <https://github.com/Isaac811>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Paddyjh/Software-Quality-Assurance-Group-10/pull/8#pullrequestreview-1956166352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARJDY65I74OX234DHIVZ6P3YZTCUHAVCNFSM6AAAAABFEEYA6OVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNJWGE3DMMZVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> <Paddyjh/Software-Quality-Assurance-Group-10/pull/8/review/1956166352@
> github.com>
>

Rosscrowley committed:
Merge branch 'develop' into feature/Task-estimation-in-Scrum

Rosscrowley committed:
Task-Estimation-In-Scrum-added-new-headings

Rosscrowley committed:
Task-Estimation-In-Scrum-added-content-for-challenges-considerations-and-techniques

Isaac811 committed:
Merge branch 'feature/Task-estimation-in-Scrum' of https://github.com/Paddyjh/Software-Quality-Assurance-Group-10 into feature/Task-estimation-in-Scrum

Isaac811 committed:
Task-Estimation-In-Scrum-added-team-velocity-and-review-adjustment

Paddyjh committed:
Task-Estimation-introduction

Added an introduction for task estimation in scrum

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+#### Sprint Retrospective
+- Time should be allocated in the sprint retrospective to address estimations; this allows the team to identify common trends of over and under-estimation.
+- All team members should be included in these discussions, as each member can enrich the discussion and offer unique insights into how to more efficiently estimate a given task.
+![sprint-retrospective-diagram](/Image_Folder/sprint-retrospective-diagram.png)
```
I think this is a good diagram that explains a lot of information, but I am unsure if it belongs in this section; we could maybe move it somewhere else, or do you guys think it is ok where it is? @Isaac811  @Rosscrowley

Paddyjh replied:
@Rosscrowley @Isaac811 I just added a diagram to address this comment. What do you guys think?

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+
+	- Challenges
+
+		- Optimism and Individual Bias: Estimators may experience optimism or personal bias which can hinder on an accurate estimation.
```
I think this could be rephrased to "which can hinder accurate estimation" or something else potentially

Rosscrowley replied:
I agree, I will rephrase. Thanks

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+
+		- Optimism and Individual Bias: Estimators may experience optimism or personal bias which can hinder on an accurate estimation.
+		- Understanding Task Complexity: Misunderstanding the task or not knowing the full breadth of the task can lead to inaccurate task estimations.
+		- Stakeholder / External pressure: Project constraints such as deadlines and budgets can influence the estimation process.
```
May be worth highlighting how they influence the process, like adding pressure or potentially discussing how this could lead teams to underestimate the length of tasks to try and achieve deadlines. I think this is discussed in the section on avoidances in "Team Velocity and Calculating Estimation below, so maybe we could just link to it?

Rosscrowley replied:
I will update this. thanks

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+		- Optimism and Individual Bias: Estimators may experience optimism or personal bias which can hinder on an accurate estimation.
+		- Understanding Task Complexity: Misunderstanding the task or not knowing the full breadth of the task can lead to inaccurate task estimations.
+		- Stakeholder / External pressure: Project constraints such as deadlines and budgets can influence the estimation process.
+		- Parkinson's Law / Changing requirements: Parkinson's law is the observation of work expanding (new requirements) to fill the time available for its completion 
```
I feel like this could be expanded or clarified more to highlight its relevance . Maybe an example or just a few more words could help.

Rosscrowley replied:
Agreed, thanks

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+#### Dynamic estimation
+- Tasks should be regularly monitored during the sprint and if progress varies significantly from estimates, this should be flagged, identified why and then taken into consideration for planning similar tasks in the future
+- Task estimation is not an exact science, and estimates should be adjusted dynamically during the sprint as new information becomes available to ensure the team is informed and deadlines can be adjusted accordingly.
+- The team should experiment with different planning techniques as mentioned previously until they find one that produces the most accuracy for their team
```
Maybe a diagram of the process of how estimations change and evolve could be useful to add here to help visualise the cyclical nature of this process.

Paddyjh replied:
Added diagram to resolve this

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+#### Calculating Velocity
+- Historical data should be used to inform the team velocity based on story points or hours completed. This ensures that estimates are more informed and accurate
+- A consistent measurement process should be used to ensure all historical data is accurate, or if measurement metrics are changed this should be taken into account when viewing historical data
+- Teams should constantly review their team velocity, taking into account both historical and any new data added since the previous calculation
```
I think a diagram could also be useful here to highlight the cyclical process a team goes through when evaluating its capacity

Paddyjh replied:
added diagram to resolve this

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,94 @@
+- A culture of learning and continuous improvement should be installed in the scrum team.
+- Discussion should be open; people should feel safe to explain why they misestimated their tasks, and others should be willing to provide feedback to them on how to improve. Individuals should embrace feedback to be more accurate in the future.
+- The team should following the following priciples in estimation
+![principles-of-good-agile-estimation](/Image_Folder/principles-of-good-agile-estimation.png)
```
This diagram is great, but it takes up a lot of space. I wonder if it is possible to address this either in markdown or by manually making the diagram smaller.

Rosscrowley replied:
As discussed in the Feature/code reviews PR, I will update the images to include a Max height of 500. Thanks

Paddyjh replied:
Brilliant, thanks!

Rosscrowley committed:
Task-Estimation-In-Scrum-changes-to-address-comments

Rosscrowley committed:
Merge branch 'feature/Task-estimation-in-Scrum' of https://github.com/Paddyjh/Software-Quality-Assurance-Group-10 into feature/Task-estimation-in-Scrum

Paddyjh committed:
Task-Estimation-added-out-task-process-diagram

I created a diagram on canva to show our recommended cyclical task estimation process.

This addresses a suggestion made on the current pull request.

Paddyjh committed:
Task-Estimation-fixed spelling-mistake-in-diagram

I fixed two spelling mistakes in our task estimation cycle diagram.

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,97 @@
+
+- There are many different approaches and techniques to estimating tasks, such as:
+
+	 - Planning Poker: Where team members estimate the effort of a task by playing a number card face down, then discussing all the estimates and coming to an agreement
```
From reading this, I'm still not entirely sure how planning poker works, do you think we could add a bit more to explain the process?

Rosscrowley committed:
Task-Estimation-In-Scrum-updating-image-format

Rosscrowley committed:
Merge branch 'feature/Task-estimation-in-Scrum' of https://github.com/Paddyjh/Software-Quality-Assurance-Group-10 into feature/Task-estimation-in-Scrum

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,97 @@
+#### Team Culture
+- A culture of learning and continuous improvement should be installed in the scrum team.
+- Discussion should be open; people should feel safe to explain why they misestimated their tasks, and others should be willing to provide feedback to them on how to improve. Individuals should embrace feedback to be more accurate in the future.
+- The team should following the following priciples in estimation
```
change the first "following" to "follow"

Rosscrowley committed:
Task-Estimation-In-Scrum-updating-image-format-2

Paddyjh reviewed (changes requested):
Two changes suggested before merging

Rosscrowley commented on Handbook/QA Handbook.md:
```diff
@@ -3,6 +3,99 @@
+  
+### Introduction  
+- Task estimation is the process a scrum team goes through outlining how long a given task or user story will take to complete. 
+    - Teams usually measure using units like story points, ideal days or hours. 
```
Given the agile nature of our team, I think we should remove mentioning measuring tasks using days or hours as this no longer fits the agile methodology. Instead, we could describe what a story point measures. Thanks

Isaac811 committed:
Task-Estimation-addressed-pr-comment

Isaac811 committed:
Task-Estimation-fixed-spelling

Paddyjh committed:
Merge branch 'develop' into feature/Task-estimation-in-Scrum

Rosscrowley reviewed (approved):
Looks good to me!

Paddyjh reviewed (approved):
Look good to me, thanks !
